frontend/src/main/java/com/google/tripmeout/frontend/servlets/PlaceParentServlet.java

```diff
@@ -0,0 +1,110 @@
+import com.google.maps.model.PlaceDetails;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PlaceParentServlet extends HttpServlet {
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private final PlaceVisitStorage placeStorage;
+  private final Gson gson;
+
+  private static final Pattern TRIP_NAME_PATTERN = Pattern.compile(".*/trips/([^/]+)/placeVisits");
+
+  private final GeoApiContext CONTEXT =
+      new GeoApiContext.Builder().apiKey("AIzaSyBbXCXC2uWv3baNmirLtqUYbFsFwCXqLV8").build();
+
+  public PlaceParentServlet(PlaceVisitStorage placeStorage, Gson gson) {
+    this.placeStorage = placeStorage;
+    this.gson = gson;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      Optional<PlaceVisitModel> place =
+          ServletUtil.extractFromRequestBody(request, gson, PlaceVisitModel.class);
+
+      String tripId = ServletUtil.matchUriOrThrowError(request, TRIP_NAME_PATTERN).group(1);
+      logger.atInfo().log("finished matching");
+
+      if (!place.isPresent()) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      PlaceDetailsRequest placeRequest =
+          new PlaceDetailsRequest(CONTEXT)
+              .placeId(place.get().placeId())
+              .fields(PlaceDetailsRequest.FieldMask.GEOMETRY_LOCATION,
+                  PlaceDetailsRequest.FieldMask.NAME);
+
+      PlaceDetails details;
+      details = placeRequest.await();
+
+      PlaceVisitModel newPlace = PlaceVisitModel.builder()
+                                     .setTripId(tripId)
+                                     .setPlaceId(place.get().placeId())
+                                     .setUserMark(PlaceVisitModel.UserMark.YES)
+                                     .setName(details.name)
+                                     .setLatitude(details.geometry.location.lat)
+                                     .setLongitude(details.geometry.location.lng)
+                                     .build();
+
+      placeStorage.updateUserMarkOrAddPlaceVisit(newPlace, PlaceVisitModel.UserMark.YES);
+      PlaceVisitModel newOrUpdatedPlace =
+          placeStorage.getPlaceVisit(tripId, place.get().placeId()).get();
+
+      response.setStatus(HttpServletResponse.SC_OK);
+      response.setContentType("application/json");
+      PrintWriter writer = response.getWriter();
+      writer.println(gson.toJson(newOrUpdatedPlace));
+      writer.flush();
+      writer.close();
+
+    } catch (IllegalArgumentException e) {
```

Comment:
This might catch unrelated exceptions and surface the wrong error to clients. See my comments in #64 about how this should be addressed in ServletUtil and here.